WIP: write data when subcycling is used

BenjaminRodenberg opened the pull request:
## Main changes of this PR

Allows to write data to preCICE when subcycling is used. This either results in a piecewise constant or linear interpolation. Possibly also in higher-order interpolation.

## Motivation and additional information

see #1171, this is an updated version of #1171 and #1523 implementing the changes introduced (especially) via #1639 and #1612.

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [x] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I sticked to C++17 features.
* [x] I sticked to CMake version 3.16.3.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

<!-- Tag people next to each point and add points for specific questions -->

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

<!-- add more questions/tasks if necessary -->

BenjaminRodenberg commented:
Some interesting findings on this PR: When running the experiments from https://github.com/BenjaminRodenberg/oscillator-example/tree/main/precice-3 with this PR and #1664 there is a clear performance difference. Just some preliminary experiments on my notebook, but for this PR we are in the order of magnitude of three minutes, while for #1664 we only need a few seconds.

BenjaminRodenberg commented:
I did a performance comparison of #1671 (this PR at c05c3c63206ecc820912d46016c1c4c10fa7da7a), #1664 (at c2b83f7b3252a30400a70ef950868a037032023a), and the current state of develop (at 49ba127e4ada06a1cbb2d81d2ac71aa7845ab665). As a basis, I used https://github.com/BenjaminRodenberg/oscillator-example/commit/05382282b52c7776585ce294fca9dcd605767fad and ran `oscillator-example/precice-3$ python3 mass-spring.py Mass-Right & python3 mass-spring.py Mass-Left`.

For develop and #1671 the experiment runs quickly, while for #1664 there is a clear performance degradation. Numerical results are all identical. There is also consistent behavior for some other tests that I ran, but I think the case above is the simplest one. Next step: Try to find the reason for the performance degradation of #1671.

davidscn commented on src/precice/impl/WriteDataContext.cpp:
```diff
@@ -31,7 +31,7 @@ void WriteDataContext::resetData(bool atEndOfWindow, bool isTimeWindowComplete)
   if (isTimeWindowComplete && hasWriteMapping()) { // manually overwrite value at beginning with value from end. Need this exception for WriteDataContext with write mapping, because CouplingScheme is not able to update _providedData, if write mapping sits between _providedData and _toData. CouplingScheme in this case only has access to _toData.
     PRECICE_ASSERT(atEndOfWindow, "isTimeWindowComplete without atEndOfWindow is forbidden!");
-    auto atEnd = _providedData->timeStepsStorage().stamples().back().sample;
+    const auto &atEnd = _providedData->timeStepsStorage().stamples().back().sample;
```
I did not read through the whole code, but considering that you are investigating performance, [this section about Eigen and `auto`](https://eigen.tuxfamily.org/dox/TopicPitfalls.html#title3) might be of interest.

BenjaminRodenberg replied:
Thank you for the hint! That's very good to know, but looks to me that the problem is somewhere else. I changed the `auto` to `Eigen::VectorXd` at several places without a observable difference in runtime. However, we actually use `auto` at many places where we access `Data::values()` (e.g. `Acceleration::applyRelaxation`).

BenjaminRodenberg commented:
I found a simple way to reproduce the performance degradation. See [b8b1779](https://github.com/BenjaminRodenberg/precice/commit/b8b17793899b7ea9c525f9e53278127aa393b11a). Here, I branched off from the latest version of #1664 and just added some communication that is similar to the communication I introduced in #1671, but without serialization and only exchanging 2 samples for any window.

BenjaminRodenberg commented:
> I found a simple way to reproduce the performance degradation. See [b8b1779](https://github.com/BenjaminRodenberg/precice/commit/b8b17793899b7ea9c525f9e53278127aa393b11a). Here, I branched off from the latest version of #1664 and just added some communication that is similar to the communication I introduced in #1671, but without serialization and only exchanging 2 samples for any window.

@uekerman and I decided to make this an issue for the future (https://github.com/precice/precice/issues/1679). Main reason: We expect that the performance degradation is less relevant for realistic cases.

BenjaminRodenberg commented:
I removed [9cdcbd2](https://github.com/precice/precice/pull/1696/commits/9cdcbd227e5cbf237254995653e112188f81c17d) and [b818e43](https://github.com/precice/precice/pull/1696/commits/b818e43e2e8b27c5edce5a268c712bf173b5b10e) from here. The goal of this PR is to provide a basic implementation for higher-order waveforms + subcycling. For a simple fixed-point iteration this also works nicely. For constant underrelaxation and more advanced acceleration schemes there is still a lot of work to do. This will be done in follow-up work packages (like #1696).

uekerman commented on src/com/SerializedStamples.hpp:
```diff
@@ -86,13 +86,17 @@ class SerializedStamples {
    */
   void serializeValues(const cplscheme::PtrCouplingData data);
 
+  void serializeValuesInitialization(const cplscheme::PtrCouplingData data);
```
Could you add here and below a quick comment explaining the difference between those methods?

BenjaminRodenberg replied:
Done. [9fb3bd7](https://github.com/precice/precice/pull/1671/commits/9fb3bd77800d65057d303b788d2100f467b10e19)

uekerman commented on src/m2n/M2N.cpp:
```diff
@@ -330,6 +346,28 @@ void M2N::receive(double &itemToReceive)
+void M2N::receive(precice::span<double> itemsToReceive)
+{
+  PRECICE_TRACE(utils::IntraComm::getRank());
+  if (not utils::IntraComm::isSecondary()) { //coupling mode
```
```suggestion
  if (not utils::IntraComm::isSecondary()) {
```
"coupling mode" as a term should be rather obsolete, I hope

BenjaminRodenberg replied:
Yes. Copy-pasted this from somewhere else. I cleaned this up via e4c706d341bff0d5156e7918fba4d53f5afccfbf.

uekerman reviewed (approved):
Looks good. I did not look in detail through the integration tests.
